### DISCUSSION TITLE
Azure Monitor Improvements

### DISCUSSION THREAD

**Discussion by joshhunt** [grafana/grafana#40555] [Data Sources]
In Grafana 8.0 we shipped a big revamp of the Azure Monitor data source to port the UI to React and added support for querying _any_ resource with Kusto, rather than just Log Analytics workspaces. Here's my thoughts on where to take Azure Monitor next.

## Quality of life fixes and feature requests

There's a number of quality of life improvements to fix some of the rough edges with the new interface we shipped in 8.0, especially around the Resource Picker:

 - [ ] https://github.com/grafana/grafana/issues/35646 - to gracefully handle large Azure accounts
 - [ ] https://github.com/grafana/grafana/issues/35643 - currently there's very little feedback when the resource picker fails, which should be addressed
 - [ ] https://github.com/grafana/grafana/issues/35648 - after selecting a resource, it's confusing when you reopen the resource picker and everything is greyed out and you cannot see what you've selected

Then there's some feature enhancements to the resource picker. These are of a lower priority, but would be really nice to get in:

 - [ ] https://github.com/grafana/grafana/issues/35655
 - [ ] https://github.com/grafana/grafana/issues/35652 - easily filter down the resources by type (or other factors like region)
 - [ ] https://github.com/grafana/grafana/issues/35651 - use ARG (probably?) to search for resources by name
 
## Big improvements

Going further, there's a bunch of bigger improvements to make to the Azure Monitor data source to align it's experience closer to the Azure Portal, some of which we have already started discussions on.

 -  [ ] https://github.com/grafana/grafana/issues/33710 - The datasource should use the resource picker for Metrics to declutter the Metrics UI. This would lay the ground for...
 - [ ] Multi-resource Metrics queries. For Azure Monitor Metrics in the Azure Portal, you can select multiple resources by saying "query all virtual machines in this resource group".
 - [ ] https://github.com/grafana/grafana/discussions/39970 - build a custom template variables picker to replace the difficult to use custom psuedo-functions the data source currently uses. @sarahzinger is already doing work on this.
 - [ ] Support Azure Resource Graph for template variable queries. I think ARG is a really powerful way to build template variables and could solve a lot of shortcomings of the current template variable experience. 

## Technical refactors

 - [ ] Remove deprecated App Insights and Insights analytics code after they were deprecated in 8.0.
 - [ ] Refactor datasource.ts and consider removing the "pseudo datasources" in favor of plain old functions.

**Comment by andresmgot** [comment]
Thanks for gathering these! Regarding the resource picker for metrics (#33710), you started that work, what's the status there?

**Reply by joshhunt** [reply]
Yeah I've started work on that, but it got parked as a part of reprioritization. I can't remember exactly where the work was up to, but i can share WIP branch...

**Reply by joshhunt** [reply]
https://github.com/grafana/grafana/pull/40887/files

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
